src/Microsoft.VisualStudio.Threading/JoinableTaskDependencyGraph.cs

```diff
@@ -167,6 +167,58 @@ internal static void OnTaskCompleted(IJoinableTaskDependent taskItem)
             taskItem.GetJoinableTaskDependentData().OnTaskCompleted();
         }
 
+        /// <summary>
+        /// Get all tasks inside the cadidate sets tasks, which are depended by one or more task in the source tasks list.
+        /// </summary>
+        /// <param name="sourceTasks">A collection of JoinableTasks represents source tasks.</param>
+        /// <param name="candidateTasks">A collection of JoinableTasks which represents cadidates.</param>
+        /// <returns>A set of tasks matching the condition.</returns>
+        internal static HashSet<JoinableTask> GetDependentTasksFromCandidates(IEnumerable<JoinableTask> sourceTasks, IEnumerable<JoinableTask> candidateTasks)
+        {
+            Requires.NotNull(sourceTasks, nameof(sourceTasks));
+            Requires.NotNull(candidateTasks, nameof(candidateTasks));
+
+            var candidates = new HashSet<JoinableTask>(candidateTasks);
+            if (candidates.Count == 0)
+            {
+                return candidates;
+            }
+
+            var results = new HashSet<JoinableTask>();
+            var visited = new HashSet<IJoinableTaskDependent>();
+
+            var queue = new Queue<IJoinableTaskDependent>();
+            foreach (JoinableTask? task in sourceTasks)
+            {
+                if (task != null && visited.Add(task))
```

Comment:
no, i tried to follow the pattern, because many 'foreach' code in the JoinableTask checked element, even the element in the collection type is not nullable. I will remove the nullable from the line 191, and leave the null check on line 193, just to make it safe.